OpenRA.Mods.Common/FileFormats/WavReader.cs

```diff
@@ -206,5 +208,99 @@ protected override bool BufferData(Stream baseStream, Queue<byte> data)
 				return ++currentBlock >= numBlocks;
 			}
 		}
+
+		// Format docs https://wiki.multimedia.cx/index.php/Microsoft_ADPCM
+		public sealed class WavStreamMsAdpcm : ReadOnlyAdapterStream
+		{
+			static readonly int[] AdaptationTable =
+			{
+				230, 230, 230, 230, 307, 409, 512, 614,
+				768, 614, 512, 409, 307, 230, 230, 230
+			};
+
+			static readonly int[] AdaptCoeff1 = { 256, 512, 0, 192, 240, 460, 392 };
+
+			static readonly int[] AdaptCoeff2 = { 0, -256, 0, 64, 0, -208, -232 };
+
+			readonly short channels;
+			readonly int blockDataSize;
+			readonly int numBlocks;
+
+			int currentBlock;
+
+			public WavStreamMsAdpcm(Stream stream, int dataSize, short blockAlign, short channels)
+				: base(stream)
+			{
+				this.channels = channels;
+				blockDataSize = blockAlign - channels * 7;
+				numBlocks = dataSize / blockAlign;
+			}
+
+			protected override bool BufferData(Stream baseStream, Queue<byte> data)
+			{
+				var bpred = new byte[channels];
+				var chanIdelta = new short[channels];
+
+				var s1 = new short[channels];
+				var s2 = new short[channels];
+
+				for (var c = 0; c < channels; c++)
+					bpred[c] = baseStream.ReadUInt8();
+
+				for (var c = 0; c < channels; c++)
+					chanIdelta[c] = baseStream.ReadInt16();
+
+				for (var c = 0; c < channels; c++)
+					s1[c] = baseStream.ReadInt16();
+
+				for (var c = 0; c < channels; c++)
+					s2[c] = WriteSample(baseStream.ReadInt16(), data);
+
+				for (var c = 0; c < channels; c++)
+					WriteSample(s1[c], data);
+
+				var channelNumber = channels > 1 ? 1 : 0;
+
+				for (var blockindx = 0; blockindx < blockDataSize; blockindx++)
+				{
+					var bytecode = baseStream.ReadUInt8();
+
+					// Decode the first nibble, this is always left channel
+					WriteSample(DecodeNibble((short)((bytecode >> 4) & 0x0F), bpred[0], ref chanIdelta[0], ref s1[0], ref s2[0]), data);
+
+					// Decode the second nibble, for stereo this will be the right channel
+					WriteSample(DecodeNibble((short)(bytecode & 0x0F), bpred[channelNumber], ref chanIdelta[channelNumber], ref s1[channelNumber], ref s2[channelNumber]), data);
+				}
+
+				return ++currentBlock >= numBlocks;
+			}
+
+			short WriteSample(short t, Queue<byte> data)
+			{
+				data.Enqueue((byte)t);
+				data.Enqueue((byte)(t >> 8));
+				return t;
+			}
+
+			// This code contains elements from libsndfile
+			short DecodeNibble(short nibble, byte bpred, ref short idelta, ref short s1, ref short s2)
+			{
+				var predict = ((s1 * AdaptCoeff1[bpred]) + (s2 * AdaptCoeff2[bpred])) >> 8;
+
+				var twosCompliment = (nibble & 0x8) > 0
+					? nibble - 0x10
+					: nibble;
+
+				s2 = s1;
+				s1 = (short)(twosCompliment * idelta + predict).Clamp(-32768, 32768);
```

Comment:
Ugh, sorry, this actually needs to be `.Clamp(-32768, 32767)`!

Clamping to +32768 introduces noticeable audio glitches in certain files (e.g. `EN-US\TDR_SFX_UNT_YESSIR1.V03_EN-US.WAV`).

LGTM after this final fix. Maybe also a good time to remove the testcase?

If anyone else wants to test this i recommend rebasing on top of #19143, adding `MegV3` to mod.yaml's `PackageFormats` list, and mounting `MUSIC.MEG` or `SFX2D_EN-US.MEG` in the packages list.